### PR TITLE
Refact clustering

### PR DESCRIPTION
**Pull request by mweiss17**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #[ISSUE_NUMBER]

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/covid19sim/frozen/clustering/naive.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+
+        Returns:
+            A tuple of the best match score (an integer in `[-1,message_uid_bit_count]`) and of
+            the matched encounter index (for internal use).
+        """
+        if message.risk_level != self.risk_level:
+            # if the cluster's risk level differs from the given encounter's risk level, there is
+            # no way to match the two, as the cluster should be updated first to that risk level,
+            # or we are looking at different users entirely
+            return -1, None
```
Comment (mweiss17) [inline src/covid19sim/frozen/clustering/naive.py]: I'm not sure about this assumption -- older messages with different risk scores should be able to be in the same cluster, right?

Comment (mweiss17) [reply src/covid19sim/frozen/clustering/naive.py]: Unless clusters are day-specific

File: src/covid19sim/frozen/clustering/naive.py [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+        if message.encounter_time in self.uids and self.uids[message.encounter_time] != message.uid:
+            # if we already have a stored uid at the new encounter's timestamp with a different
+            # uid, there's no way this message can be merged into this cluster
+            return -1, None
+        best_match = (-1, None)
+        assert len(self.messages), "... a cluster cannot exist without past encounters?"
```
Comment (mweiss17) [inline src/covid19sim/frozen/clustering/naive.py]: we should list the assumptions we're making about the input data.

File: src/covid19sim/frozen/clustering/naive.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        for encounter_idx, old_encounter in enumerate(self.messages):
+            match_score = mu.find_encounter_match_score(
+                # TODO: what happens if the received message is actually late, and we have an
+                #       'old message' that is more recent? (match scorer will throw)
```
Comment (mweiss17) [inline src/covid19sim/frozen/clustering/naive.py]: And have better exception handling. Basically, we want to ensure the clustering simply degrades if shit is whack, not breaks.

Comment (mweiss17) [reply src/covid19sim/frozen/clustering/naive.py]: degrades and logs stuff.

File: src/covid19sim/frozen/clustering/naive.py (hunk 1 above)
Comment (plstcharles) [reply src/covid19sim/frozen/clustering/naive.py]: two encounters can only be from the same person and coexist in the same cluster if they both share the same risk level; if that is not the case, then either it was two persons (and the cluster should rightfully be split), or we are in the process of updating messages and only received a partial update (in which case, we will split and partially update)

in other words, multiple encounters can coexist in a cluster only if they have all been updated to the same risk level

File: src/covid19sim/frozen/clustering/naive.py (hunk 2 above)
Comment (plstcharles) [reply src/covid19sim/frozen/clustering/naive.py]: good point; in this case, the assumption is only specific to the clustering logic, and means that we cannot have an "empty cluster", as the split/merge operations should never result in that

**Comment by plstcharles** [comment]
the refactoring seems complete now, but there is still much to do in terms of evaluation; I will now setup a dataset for I/O testing, and update @mweiss17 's original plotting script to support the new code

**Comment by plstcharles** [comment]
note: this is definitely not ready for merging, as there are still cases that throw once integrated in the simulation (hard to know who/what/where to fix)

**Review by plstcharles — APPROVED** [review]

**Pull request merged**